src/hotspot/share/opto/divnode.cpp

```diff
@@ -504,6 +504,92 @@ Node* unsigned_div_ideal(PhaseGVN* phase, bool can_reshape, Node* div) {
   return nullptr;
 }
 
+template<typename IntegerType>
+static const IntegerType* compute_generic_div_type(const IntegerType* i1, const IntegerType* i2, int widen) {
+  typedef typename IntegerType::NativeType NativeType;
+  assert(!i2->is_con() || i2->get_con() != 0, "Can't handle zero constant divisor");
+
+  // Case A: divisor range spans zero (i2->_lo < 0 < i2->_hi)
+  // We split into two subproblems to avoid division by 0:
+  //   - negative part: [i2->_lo, −1]
+  //   - positive part: [1, i2->_hi]
+  // Then we union the results by taking the min of all lower‐bounds and
+  // the max of all upper‐bounds from the two halves.
+  if (i2->_lo < 0 && i2->_hi > 0) {
+    // Handle negative part of the divisor range
+    const IntegerType* neg_part = compute_generic_div_type(i1, IntegerType::make(i2->_lo, -1, widen), widen);
+    // Handle positive part of the divisor range
+    const IntegerType* pos_part = compute_generic_div_type(i1, IntegerType::make(1, i2->_hi, widen), widen);
+    // Merge results
+    NativeType new_lo = MIN2(neg_part->_lo, pos_part->_lo);
+    NativeType new_hi = MAX2(neg_part->_hi, pos_part->_hi);
+    assert(new_hi >= new_lo, "sanity");
+    return IntegerType::make(new_lo, new_hi, widen);
+  }
+
+  // Case B: divisor range does NOT span zero.
+  // Here i2 is entirely negative or entirely positive.
+  // Let d_min and d_max be the nonzero endpoints of i2.
```

Comment:
Seems you define `d_min` and `d_max` here, but you don't use them anywhere. You should probably use names here that you will use further down.